topics/03_file_handling_and_regex/HW_03_file_handling_and_regex.py

```diff
@@ -0,0 +1,164 @@
+# 03_file_handling_and_regex -------------------------------------------------------------------------------------------
+# 1 [TI] Explain the difference between read(), readline(), and readlines(). In what situations would you use each?
+# read() is best for smaller files or when you need to manipulate the whole text at once.
+# readline() is optimal for reading a file line by line, particularly with large files.
+# readlines() suits scenarios where you need the entire file read into a list for easy manipulation of line-based data within a list context, but be mindful of memory usage with larger files.
+# 2 [TI] How does the with statement help in file handling, and why is it preferred over manually opening and closing files?
+# with statement provides a much more readable, cleaner,
+# and error-free way to handle files in Python,
+# especially when compared to the traditional approach of manually opening
+# and closing files. It makes your file handling safer and more consistent.
+import os
+file_dir = os.path.dirname(os.path.realpath('C:/Users/Olena_Smolianchuk/Downloads/'))
+print(file_dir)
+file_name = os.path.join(file_dir, 'Downloads/config.ini')
+# 3 Write a script that reads a file test_data.txt and prints each line.
+with open('test_data.txt', "r") as file:
+    for line in file:
+        print(line)
+# 4 Modify the script from the previous exercise to append a new line to the file.
+with open('test_data.txt', "a") as file:
+    file.write ("This text should be added as new line to file")
+# 5 Write a script that reads test_data.txt and counts the number of lines and words in the file. Print both counts at the end.
+file_path = 'test_data.txt'
+line_count = 0
+word_count = 0
+with open(file_path, 'r') as file:
+    for line in file:
+        line_count += 1
+        words = line.split()
+        word_count += len(words)
+print(f"Total lines: {line_count}")
+print(f"Total words: {word_count}")
+# 6 Write a script that replaces a specific word (e.g., "error") in a file logs.txt with the word "fixed" and saves the changes.
+with open('logs.txt', 'r') as file:
+    data = file.readlines()
+
+with open('logs.txt', 'w') as file:
+    for line in data:
+        new_line = line.replace('error', 'fixed')
+        file.write(new_line)
+# 7 Write a script that tries to read a file data.txt, handles the exception if the file does not exist, and prompts the user to create the file if it's missing.
+def read_file():
+    try:
+        with open('data.txt', 'r') as file:
+            data = file.read()
+            print("File content:")

```

Comment:
printing can be done in one line using string formatting:
print(f"File content:\n{data}")
\n is for the next line